groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+        reinterpret_cast<TYPE *>(BSLS_UTIL_ADDRESSOF(d_union)),
+        AllocBase::mechanism());
+    d_type = BSLSTL_VARIANT_INDEX_OF(TYPE, variant);
+}
+
+template <class HEAD, class... TAIL>
+template <class TYPE, class ARG>
+inline
+variant<HEAD, TAIL...>::variant(
+                              bsl::allocator_arg_t,
+                              allocator_type                         allocator,
+                              bsl::in_place_type_t<TYPE>,
+                              BSLS_COMPILERFEATURES_FORWARD_REF(ARG) arg)
+: AllocBase(allocator)
+, d_type(bsl::variant_npos)
+{
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<TYPE *>(BSLS_UTIL_ADDRESSOF(d_union)),
+        AllocBase::mechanism(),
+        BSLS_COMPILERFEATURES_FORWARD(ARG, arg));
+    d_type = BSLSTL_VARIANT_INDEX_OF(TYPE, variant);
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX>
+inline
+variant<HEAD, TAIL...>::variant(bsl::allocator_arg_t,
+                                allocator_type               allocator,
+                                bsl::in_place_index_t<INDEX>)
+: AllocBase(allocator)
+, d_type(bsl::variant_npos)
+{
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<typename variant_alternative<INDEX, variant>::type *>(
+            BSLS_UTIL_ADDRESSOF(d_union)),
+        AllocBase::mechanism());
+    d_type = INDEX;
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX, class ARG>
+inline
+variant<HEAD, TAIL...>::variant(
+                              bsl::allocator_arg_t,
+                              allocator_type                         allocator,
+                              bsl::in_place_index_t<INDEX>,
+                              BSLS_COMPILERFEATURES_FORWARD_REF(ARG) arg)
+: AllocBase(allocator)
+, d_type(bsl::variant_npos)
+{
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<typename variant_alternative<INDEX, variant>::type *>(
+            BSLS_UTIL_ADDRESSOF(d_union)),
+        AllocBase::mechanism(),
+        BSLS_COMPILERFEATURES_FORWARD(ARG, arg));
+    d_type = INDEX;
+}
+#endif
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD, class... TAIL>
+variant<HEAD, TAIL...>::~variant()
+{
+    reset();
+}
+// MANIPULATORS
+template <class HEAD, class... TAIL>
+void variant<HEAD, TAIL...>::reset() BSLS_KEYWORD_NOEXCEPT
+{
+    if (d_type != bsl::variant_npos) {
+        BloombergLP::bslstl::Variant_DestructorVisitor destructor;
+        bsl::visit(destructor, *this);
+    }
+    d_type = bsl::variant_npos;
+}
+
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+
+template <class HEAD, class... TAIL>
+template <class TYPE, class... ARGS>
+TYPE& variant<HEAD, TAIL...>::emplace(ARGS&&... args)
+{
+    BSLMF_ASSERT(
+        (BloombergLP::bslstl::
+             Variant_HasUniqueType<TYPE, variant<HEAD, TAIL...> >::value ==
+         true));
```

Comment:
```suggestion
    static_assert(
               BloombergLP::bslstl::
                   Variant_HasUniqueType<TYPE, variant<HEAD, TAIL...> >::value,
               "Type is not a unique alternative in the variant");              
```
You used `static_assert` below, so no reason not to use it here. This is the "full" implementation after all.
(I think we might even be able to not put a space between the closing angle brackets?)

Reply:
done